app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/zetime/ZeTimeDeviceSupport.java

```diff
@@ -0,0 +1,1223 @@
+                    msgpartlength += maxMsgLength;
+                }
+                builder.write(writeCharacteristic, msgpart);
+            }while(msgpartlength < msg.length);
+        } else
+        {
+            builder.write(writeCharacteristic, msg);
+        }
+        builder.write(ackCharacteristic, new byte[]{ZeTimeConstants.CMD_ACK_WRITE});
+    }
+
+    private void handleMusicControl(byte[] musicControlMsg)
+    {
+        if(musicControlMsg[2] == ZeTimeConstants.CMD_SEND) {
+            switch (musicControlMsg[5]) {
+                case 0: // play current song
+                    musicCmd.event = GBDeviceEventMusicControl.Event.PLAY;
+                    break;
+                case 1: // pause current song
+                    musicCmd.event = GBDeviceEventMusicControl.Event.PAUSE;
+                    break;
+                case 2: // skip to previous song
+                    musicCmd.event = GBDeviceEventMusicControl.Event.PREVIOUS;
+                    break;
+                case 3: // skip to next song
+                    musicCmd.event = GBDeviceEventMusicControl.Event.NEXT;
+                    break;
+                case 4: // change volume
+                    if (musicControlMsg[6] > volume) {
+                        musicCmd.event = GBDeviceEventMusicControl.Event.VOLUMEUP;
+                        if(volume < 90) {
+                            volume += 10;
+                        }
+                    } else {
+                        musicCmd.event = GBDeviceEventMusicControl.Event.VOLUMEDOWN;
+                        if(volume > 10) {
+                            volume -= 10;
+                        }
+                    }
+                    try {
+                        TransactionBuilder builder = performInitialized("replyMusicVolume");
+                        replyMsgToWatch(builder, new byte[]{ZeTimeConstants.CMD_PREAMBLE,
+                                ZeTimeConstants.CMD_MUSIC_CONTROL,
+                                ZeTimeConstants.CMD_REQUEST_RESPOND,
+                                0x02,
+                                0x00,
+                                0x02,
+                                volume,
+                                ZeTimeConstants.CMD_END});
+                        performConnected(builder.getTransaction());
+                    } catch (IOException e) {
+                        GB.toast(getContext(), "Error reply the music volume: " + e.getLocalizedMessage(), Toast.LENGTH_LONG, GB.ERROR);
+                    }
+                    break;
+            }
+            handleGBDeviceEvent(musicCmd);
+        } else {
+            if (music != null) {
+                music[2] = ZeTimeConstants.CMD_REQUEST_RESPOND;
+                try {
+                    TransactionBuilder builder = performInitialized("replyMusicState");
+                    replyMsgToWatch(builder, music);
+                    performConnected(builder.getTransaction());
```

Comment:
I don't understand the pattern
```
performInitialized()
doStuff()
performConnected()
```

Probably the performXX() methods lack information, so here's an explalation:
- at first, the device is becoming connected (BT-level)
- most devices usually need some initialization like setting the time, etc. after connection
- after that, the device is 'initialized'
- performInitialized() will allow you to write to the device in a way that will make sure that the writes will only happen *after* the connection to the device has been established *and* the initialization has successfully finished
- afterwards, you need to actually perform the writes by calling `builder.queue(getQueue())`. This will return immediately -- the rest will happen asynchronously via the queue.

So the `performConnected()` call should typically not be necessary.

Reply:
Had seen this pattern been used by other devices, but it might be that this was old or not right.